Add a PORO for creating tax adjustments

mamhoff opened the pull request:
Adds a `Spree::Tax::ItemAdjuster`, which adjusts an item, and a `Spree::Tax::OrderAdjuster`, which iterates over line items and shipments in an order and adjusts those. 

This PR is a prerequisite for any further taxation work.

mamhoff commented:
Force-pushed with no changes because of spurious failure in admin shipment spec.

jhawthorn commented:
This does seem like a good target for extraction to a PORO, but the handling of the two cases (receiving either an item or an order) is a little awkward. The two cases share very little code.

Maybe we could adjust the signature of the adjuster? Brainstorming a little:

```
class Adjuster
  def initialize(order)
    @order = order
    @tax_zone = order.tax_zone
  end

  def adjust_order
    # adjust all items in the order
  end

  def adjust_item(item)
    # adjust just the item
  end
end
```

tvdeyen commented:
One adjuster for two different things? It seams the only thing they share is the order tax zone. 

Why not an `OrderAdjuster` and a `LineItemAdjuster`, both knowing the order tax zone but having separated concerns?

Sandi Metz is calling! Excuse me ;)

tvdeyen commented:
There is too much knowledge of the two possible callers inside one class. Case statements on class types, then methods named after the caller, etc. 

Two separate adjusters are much more flexible, more future safe and less error prone. If we use POROs than please let us use clean object oriented design. 

Sorry, reading too much Sandi Metz lately. 

Just my 0.02

mamhoff commented:
Thank you for chipping in! 

@jhawthorn Actually, they do share code, as the order adjuster doesn't actually adjust the order, but all the items (line items and shipments) contained within it. 

The current stuff on `TaxRate` shaves some AR requests by only loading the `potentially_applicable` rates once per order - I would like that efficiency to stay, which is why I didn't separate the two into their own classes. @tvdeyen I hear you though. Possibly, though, I can have the best of both worlds by making the `LineItemAdjuster` a Singleton and cache things like the tax rates in an instance variable. 

At the moment, `Spree::TaxRate.adjust` is a method that is executed for side effects (create adjustments, update `pre_tax_amount`), which isn't super transparent. Ideally, I'd like something to this effect: 

```
class Spree::LineItem
  has_many :adjustments, ... , autosave: true
  before_save :adjust

  private 
  
  # builds tax adjustments and sets pre tax amount, but doesn't actually save
  def adjust
    self = Spree::Tax::ItemAdjuster.new(self).adjust!
  end
end
```

mamhoff commented:
Changed these to single responsibility objects - I do think we can deal with the efficiency after we've defined the interfaces.

tvdeyen commented on core/app/models/spree/order.rb:
```diff
@@ -324,10 +324,7 @@ def line_item_options_match(line_item, options)
-      order_tax_zone = self.tax_zone
-      Spree::TaxRate.adjust(order_tax_zone, line_items)
-      Spree::TaxRate.adjust(order_tax_zone, shipments) if shipments.any?
+      Spree::Tax::Adjuster.new(self).adjust!
```
Don't you want to call `OrderAdjuster` here?

mamhoff replied:
Absolutely. Thanks for the heads-up! :100:

mamhoff commented:
Wheew! Green! Removing `TaxRate.adjust`, which is a monster, has not been that hard after all. I'm excited: I like the API, and because of some data sharing between the `ItemAdjuster` and the `OrderAdjuster` the expensive `TaxRate.match` call is only done once. 

This is still lacking feedback. Once that's in, I'd start putting in unit tests and moving the Spree::TaxRate into some kind of `taxation_integration_spec.rb` file. 

Please please, this isn't ready for merge, but definitely ready for feedback (although you do have a couple of days as I'll be out until the weekend from tomorrow).

mamhoff commented:
Rebased on master, updated commit message on second commit. Feedback still welcome. 

Btw: With #530 in, I'll take a go at removing `TaxRate.match`, too.

mamhoff commented:
These last ones failed because of the Rails update (https://github.com/solidusio/solidus/pull/736). Rebasing on master now, and adding a couple of performance fixes as well.

mamhoff commented:
The build failed because of phantomjs crashing :(

mamhoff commented:
Force-pushed to get a green build. Otherwise: This is ready for review. 

I can think of more stuff to do, but the scope of this PR is already quite large, so I'll stop for now. :)

peterberkenbosch commented:
:+1: Nice @mamhoff

mamhoff commented:
Damn. The shipping rates also have VAT, and they're not price-adjusted yet.

tvdeyen commented:
I hopefully can spent some review time later this evening (CET)

mamhoff commented:
This is really long, and hard to review. I'll split this PR in two: One that just adds the POROs and uses #530 to look up zones (that's all commits up to a68d9c8), and one that actually implements MOSS with that. 

The one that actually implements MOSS will also have to deal with shipping rates, somehow.

tvdeyen commented:
Alright

tvdeyen commented on core/app/models/spree/tax/item_adjuster.rb:
```diff
@@ -0,0 +1,32 @@
+
+      def initialize(item, options = {})
+        @item = item
+        @order = @item.order
```
`@item` and `@order` should be `attr_reader`s, so the `TaxHelper` `module` and potential sub classes don't rely on reading instance variables.

tvdeyen commented on core/app/models/spree/tax/item_adjuster.rb:
```diff
@@ -0,0 +1,32 @@
+      def adjust!
+        return unless order_tax_zone
+        # Using .destroy_all to make sure callbacks fire
+        @item.adjustments.tax.destroy_all
```
after introduced we should use the `item` `attr_reader` here and not use instance variables.

tvdeyen commented on core/app/models/spree/tax/item_adjuster.rb:
```diff
@@ -0,0 +1,32 @@
+      private
+
+      def rates_for_item
+        @rates ||= applicable_rates.select { |rate| rate.tax_category == @item.tax_category }
```
I find it good practice to name caching instance variables after the methods name and use an underscore to make it more visible, that this is not a normal instance variable. So, the name should be `@_rates_for_item` IMO. 

And we should also use the `item` `attr_reader` here.

tvdeyen commented on core/app/models/spree/tax/order_adjuster.rb:
```diff
@@ -0,0 +1,19 @@
+      include TaxHelpers
+
+      def initialize(order)
+        @order = order
```
Same as in item adjuster. Use a `attr_reader` for `order`

tvdeyen commented on core/app/models/spree/tax/tax_helpers.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+
+      def order_tax_zone
+        @order_tax_zone ||= @order.tax_zone
```
This is exactly why we should use the `order` `attr_reader`. The `module` shouldn't know anything about an instance variable set in the including class.

tvdeyen replied:
also caching variable name

tvdeyen commented on core/app/models/spree/tax/tax_helpers.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+
+      def rates_for_default_zone
+        @rates_for_default_zone ||= Spree::TaxRate.for_zone(Spree::Zone.default_tax)
```
Here the same rule for naming caching instance variables applies

tvdeyen commented on core/spec/models/spree/tax/item_adjuster_spec.rb:
```diff
@@ -0,0 +1,29 @@
+  describe 'initialization' do
+
+    it 'sets @order to adjustable.order' do
+      expect(adjuster.instance_variable_get(:@order)).to eq(item.order)
```
A prime example why we need the `order` `attr_reader`

tvdeyen replied:
```ruby
expect(adjuster.order).to eq(item.order)
```

:heart_eyes:

mamhoff commented:
Did the naming changes, and also added some more tests to the adjusters.

mamhoff commented:
First commit was failing because of a naming change, rebased again.

gmacdougall commented on core/app/models/spree/tax/item_adjuster.rb:
```diff
@@ -0,0 +1,32 @@
+        @item = item
+        @order = @item.order
+        # set the instance variable so `TaxRate.match` is only called when necessary
+        @rates_for_order_zone = options[:rates_for_order_zone]
```
I'm not clear where this is being used.

mamhoff replied:
These are used in private methods of the included module `TaxHelpers`, in order to run the database calls that only have to run once per order only once per order: https://github.com/magiclabs/solidus/blob/add-tax-adjuster/core/app/models/spree/tax/tax_helpers.rb#L28-L38.

gmacdougall commented on core/app/models/spree/tax/item_adjuster.rb:
```diff
@@ -0,0 +1,32 @@
+        @rates_for_order_zone = options[:rates_for_order_zone]
+      end
+
+      def adjust!
```
Could we add some YARD here to explain the return values of this method.

I also tend to to not have a method ending in a bang without a corresponding non-bang method.  See http://dablog.rubypal.com/2007/8/15/bang-methods-or-danger-will-rubyist

mamhoff replied:
I can very well imagine a future non-bang `adjust` method: One that will not change the order, and not create adjustments, but just `build` them. That would be great to reduce database calls when updating items and orders. Alas, we're not there yet, but I'd like to keep this method's name the same at that point. 

YARD I can absolutely add, as this method returns the newly created adjustments (making it somewhat easier to test than `Spree::TaxRate.adjust`).

gmacdougall commented on core/app/models/spree/tax/order_adjuster.rb:
```diff
@@ -0,0 +1,21 @@
+      def adjust!
+        return unless order_tax_zone
+
+        (order.line_items + order.shipments).each do |item|
```
Is this a concept we should have on the order model? Some sort of method that returns all things that can be adjusted?

mamhoff replied:
We could, potentially. I shied away of adding more methods to `Spree::Order`. I'd do that on a later commit though once we actually see it DRYing away other things, namely when I have to write a method that does the same thing for the shipping rates in a shipment (https://github.com/solidusio/solidus/pull/759).

gmacdougall commented on core/app/models/spree/zone.rb:
```diff
@@ -58,6 +58,8 @@ def self.match(address)
     # current zone, if it's a state zone. If the passed-in zone has members, it
     # will also be in the result set.
     def self.with_shared_members(zone)
+      return none unless zone
```
Does this work?

jhawthorn replied:
Looks like it should to me. This is a class method so it will return the `Spree::Zone.none` null object relation.

mamhoff replied:
What John said. However, medium term that entire method goes away when #783 gets applied.

gmacdougall replied:
I had forgotten that `none` is an ActiveRecord method. Looked for our own implementation and didn't see one, and got confused. Sorry spreading the confusion.

jhawthorn commented:
Looks good to me. :+1:

One thought, should we keep `TaxRate.adjust` as a deprecated method that would use the PORO?

mamhoff commented:
Well, the method signatures are a bit different. I could use just the item adjuster, but then that method will have abysmal performance. I'd rather have it entirely gone. However, if there's another person saying: Deprecate it and be nice about it - I'll do it, along with the requested YARD docs tomorrow.

tvdeyen commented:
:icecream:

timmyArch commented:
Nice ;)
